### PR TITLE
Events for web frameworks

### PR DESCRIPTION
**Pull request by nizox**
## Description
On the road to decouple the framework integrations and the tracer code, this PR starts small by introducing an event interface. The web framework integrations are modified to replace the `trace_utils.set_http_meta` function calls by Web request & response events. The Tracer also adds listeners to set HTTP tags using the event interface.

## Checklist
- [ ] Added to the correct milestone.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

File: ddtrace/events.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
+from typing import List
```
Comment (Kyle-Verhoog) [inline ddtrace/events.py:1]: Let's keep this API private for now `_events.py` is fine with me :+1:

File: ddtrace/events.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+from typing import Union
+
+from ._hooks import Hooks
+
+
+if TYPE_CHECKING:
+    from ddtrace.span import Span
+
+
+HTTP_HOOKS = Hooks()
```
Comment (Kyle-Verhoog) [inline ddtrace/events.py:14]:
```suggestion
_HTTP_HOOKS = Hooks()
```

**Review by Kyle-Verhoog — COMMENTED** [review]
did a quick pass - looks great! just initial concerns about api visibility 👍

File: ddtrace/contrib/pylons/middleware.py [hunk 3]
```diff
@@ -57,7 +58,14 @@ def __call__(self, environ, start_response):
             # set analytics sample rate with global config enabled
             span.set_tag(ANALYTICS_SAMPLE_RATE_KEY, ddconfig.pylons.get_analytics_sample_rate(use_global_config=True))
 
-            trace_utils.set_http_meta(span, ddconfig.pylons, request_headers=request.headers)
```
Comment (P403n1x87) [inline ddtrace/contrib/pylons/middleware.py:60]: So is this a first step towards an event-based approach to tracing? 🙂

File: ddtrace/_events.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    from ddtrace.span import Span
+
+
+_HTTP_HOOKS = Hooks()
+
+REQUEST = "request"
+RESPONSE = "response"
+
+
+def emit_http_request(
```
Comment (jd) [inline ddtrace/_events.py:20]: What about having something more generic such as:
`emit_event(event)`

where event would be a class such as `HTTPEvent` like:
```
class Event:
    span: Span
    integration: str

class HTTPEvent(Event):
    method: str
    url: str
    …
```

That'd make the system reusable for other events.

You can then use a `register_event(event_cls)` which register function for particular event types.

Comment (nizox) [reply ddtrace/_events.py]: I like it! It also comes with a nice to have feature: we can now add validators to attributes for unit testing 🥳

File: ddtrace/_events.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+@attr.s(frozen=True)
+class IntegrationEvent(object):
+    """
+    An IntegrationEvent is emitted by an integration (e.g. the flask framework integration)
+    and is linked to a span.
+    """
+    span = attr.ib(type=Span)
+    integration = attr.ib(type=str)
+
+    def emit(self):
```
Comment (jd) [inline ddtrace/_events.py:27]: We definitely want typing in all this file :)

File: ddtrace/_events.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+
+@attr.s(frozen=True)
+class WebRequest(IntegrationEvent):
+    """
+    The WebRequest event is emitted by web framework integrations before the WebResponse event.
+    """
+    method = attr.ib(type=str)
+    url = attr.ib(type=str)
+    headers = attr.ib(type=Mapping[str, str], factory=dict)
```
Comment (jd) [inline ddtrace/_events.py:58]: I'm going to be picky here, but the type is wrong AFAIK.
You can have multiple values for a header, so `[str, str]` is a limited approach.
I'd rather be `Mapping[str, Iterable[str]]` or something like that.

File: ddtrace/_events.py (hunk 5 above)
Comment (nizox) [reply ddtrace/_events.py:27]: Sure! I will add more types & unit tests for this file.

File: ddtrace/_events.py (hunk 6 above)
Comment (nizox) [reply ddtrace/_events.py:58]: 💯 agreed! I will apply the changes.

File: ddtrace/contrib/pylons/middleware.py (hunk 3 above)
Comment (Kyle-Verhoog) [reply ddtrace/contrib/pylons/middleware.py:60]: yes!!!

**Review by brettlangdon — CHANGES_REQUESTED** [review]
I like it, but I would like to see performance tests before we can merge this. Probably starting with some microbenchmarks for our hooks implementation would be good, and then before/after impact on some of the web frameworks would be good.

Especially testing if we have more than 1 hook registered as well.

**Pull request closed without merging**